src/opossum_lib/file_generation.py

```diff
@@ -1,89 +1,159 @@
-    metadata: Dict[str, str]
-    resources: Dict[str, Any]
-    externalAttributions: Dict
-    resourcesToAttributions: Dict
-    attributionBreakpoints: List[str]
 
 
 def write_dict_to_file(
     opossum_information: OpossumInformation, file_path: Path
 ) -> None:
     with file_path.open("w") as out:
-        json.dump(opossum_information, out, indent=4)
+        json.dump(
+            asdict(opossum_information, dict_factory=dicts_without_none), out, indent=4
+        )
 
 
-def generate_json_file_from_tree(
-    tree: DiGraph,
-) -> OpossumInformation:
-    doc_name = tree.nodes["SPDXRef-DOCUMENT"]["element"].name
+def dicts_without_none(x: List[Tuple[str, Any]]) -> Dict[Any, Any]:
+    return {k: v for (k, v) in x if v is not None}
+
+
+def generate_json_file_from_tree(tree: DiGraph) -> OpossumInformation:
+    metadata = create_metadata(tree)
+    resources: Dict[str, Any] = dict()
+    resources_to_attributions: Dict[str, List[str]] = dict()
+    external_attributions: Dict[str, OpossumPackage] = dict()
+    attribution_breakpoints = []
 
-    opossum_information: OpossumInformation = {
-        "metadata": {
-            "projectId": "tools-python-opossum-crossover",
-            "projectTitle": doc_name,
-            "fileCreationDate": datetime.datetime.utcnow().isoformat(),
-        },
-        "resources": {},
-        "externalAttributions": {},
-        "resourcesToAttributions": {},
-        "attributionBreakpoints": [],
-    }
-
-    for connected_set in weakly_connected_components(tree):
-        connected_subgraph = tree.subgraph(connected_set).copy()
+    external_attributions["file-identifier"] = OpossumPackage(source=SourceInfo("File"))
+    external_attributions["package-identifier"] = OpossumPackage(
+        source=SourceInfo("Package")
+    )
+    external_attributions["snippet-identifier"] = OpossumPackage(
+        source=SourceInfo("Snippet")
+    )
+
+    for connected_subgraph in _weakly_connected_component_sub_graphs(tree):
         source = _get_source_for_graph_traversal(connected_subgraph)
-        if source is None:  # from development not possible, how to deal with this?
+        if source is None:
             raise RuntimeError(
                 "A tree should always have a node without incoming edge."
             )
-
-        opossum_information["resources"][source] = tree_to_dict(
-            connected_subgraph, source
-        )
-        opossum_information["attributionBreakpoints"].extend(
-            get_breakpoints(connected_subgraph, source)
+        resources_list = []
+        for node in connected_subgraph.nodes():
+            path: List[str] = shortest_path(connected_subgraph, source, node)
+            path_with_labels: List[str] = _replace_node_ids_with_labels(
+                path, connected_subgraph
+            )
+            file_path: str = _create_file_path_from_graph_path(connected_subgraph, path)
+
+            resources_list.append(_create_nested_dict(path_with_labels))
```

Comment:
Just wanted to write down one idea that just came to my mind how you could maybe circumvent the merge_nested_dict-stuff:
Instead of having this `resources_list` list of dicts, you could have a `resources_tree` DiGraph (completely independent of all other graphs here in the code) that you use to incrementally build a tree that has the path component strings as nodes.
Instead of doing `resources_list.append(_create_nested_dict(path_with_labels))`, you would then have a simple function like `add_path_to_resource_tree(resource_tree: Digraph, path: List[str]) -> None` that updates the tree.

Then in the end you could still call the existing `tree_to_dict` function to get the dict.

Reply:
I just had some fun playing around: Actually, this small class should do everything you need:
```

@dataclass(frozen=True)
class Resource:
    children: Dict[str, Resource] = field(default_factory=dict)

    def add_path(self, path: List[str]) -> None:
        if len(path) == 0:
            return
        first, rest = path[0], path[1:]
        if first not in self.children:
            self.children[first] = Resource()
        self.children[first].add_path(rest)

    def to_dict(self):
        if len(self.children) == 0:
            return 1
        else:
            return {name: resource.to_dict() for name, resource in self.children.items()}

def test_resource():
    x = Resource()
    x.add_path(["a", "b", "c"])
    x.add_path(["a", "d"])
    print(x.to_dict())  # prints: {'a': {'b': {'c': 1}, 'd': 1}}
```